### PR TITLE
Set up Supabase and connect it with Slack Bolt

### PR DESCRIPTION
**Pull request by hiroshinishio**
## Issue

So far, we have not been connected to the DB; we have substituted storing and retrieving to a pseudo-local DB for development provided by Slack.

## Solution

We linked to DB. We wondered whether to use SQL or NoSQL for DB, but we adopted Supabase, which made SQL Postgres easier.

The data exchanged regarding the installation of the Slack app is returned as a JSON object, so I thought MongoDB, a NoSQL that can store the data as it is, would be fine, but @theanhmx recommended it, and I thought it would be a good way to learn Postgres again. I decided to use Supabase, which is a legitimate SQL system.

### Install it, use it, and uninstall it

https://www.loom.com/share/f570617b274147d7977aef5bac010cf2

### Supabase's production & development environment are ready

image

## Caveats

@huitzlopochtli Please take a look at these below.

1. Unsure if table definitions are correct in the first place
2. Primary key type was set to type text, but I'm worried because I don't think it's best practice
3. I Created the table directly from the Editor, so I'm not sure if it's correct that the script doesn't keep a history of the table. I committed the script so that it could be Git managed in the meantime, but I'm not sure that's right either.
4. I made the RLS settings virtually meaningless, so I'll have to do it right, at another time.

### Slack Bolt for JavaScript

- [What is Installation Object](https://slack.dev/bolt-js/concepts#installation-object)
- [How to not use stateSecret](https://slack.dev/bolt-js/concepts#org-wide-installation)
- [How to allow direct installation for users](https://slack.dev/bolt-js/concepts#installing-your-app)
- [How to use installerOptions](https://github.com/slackapi/bolt-js/blob/5b4d9ceb65e6bf5cf29dfa58268ea248e5466bfb/examples/oauth/app.js#L58-L64)
- [Why is 'deleteInstallation' not triggered](https://github.com/slackapi/bolt-js/issues/1203)
- [Using the Web API](https://slack.dev/bolt-js/concepts#web-api)
- [Web API の使用](https://slack.dev/bolt-js/ja-jp/concepts#web-api)
- [Slack Bolt で簡単に複数ワークスペースにインストールできるアプリを開発しよう！（TypeScript, Lambda, Serverless Framework によるサンプルコード付き）](https://qiita.com/irongineer/items/9f0b997bbf62ae0d0e06)
- [Slack アプリフレームワーク Bolt for JavaScript を Cloud Functions for Firebase で動かそう](https://qiita.com/seratch/items/3e44f58a7c8ab41f21f0)
- [Bolt を使って Slack でヘルプデスクチームとのやりとりを効率化するアプリをつくってみよう](https://qiita.com/seratch/items/9b5666479ad9f602e5ee)

### Supabase

- [Install Supabase CLI](https://supabase.com/docs/guides/resources/supabase-cli)
- [Supabase CLI Intro](https://supabase.com/docs/reference/cli/introduction)
- [Data Types such as bigint](https://supabase.com/docs/guides/database/tables#data-types)
- [Primary Keys](https://supabase.com/docs/guides/database/tables#primary-keys)
- [Row Level Security](https://supabase.com/docs/guides/auth/row-level-security)

### Supabase Console

- [Team Setting](https://app.supabase.com/org/bncihrsdionehlgsazzn/settings)
- [App Setting](https://app.supabase.com/project/zpqebusroccctyubbasr/settings/general)
- [SQL Editor](https://app.supabase.com/project/zpqebusroccctyubbasr/sql)
- [Table Editor](https://app.supabase.com/project/zpqebusroccctyubbasr/editor)

### Supabase JS Library

- [Fetch Data](https://supabase.com/docs/reference/javascript/select)
- [Upsert Data](https://supabase.com/docs/reference/javascript/upsert)
- [Delete Data](https://supabase.com/docs/reference/javascript/delete)
- [eq() for using filters](https://supabase.com/docs/reference/javascript/using-filters)
- [single() for getting 1 row](https://supabase.com/docs/reference/javascript/single)

### Postgres

- [PostgreSQL Upsert Using INSERT ON CONFLICT statement](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/)
- [CREATE TABLE](https://www.postgresql.org/docs/current/sql-createtable.html)
- [DROP TABLE IF EXISTS](https://www.postgresql.org/docs/current/sql-droptable.html)
- [UUID type](https://www.postgresql.org/docs/current/datatype-uuid.html)
- [text[] type](https://www.postgresql.org/docs/current/arrays.html)
- [Row Level Security](https://www.postgresql.org/docs/current/ddl-rowsecurity.html)
- [CREATE POLICY](https://www.postgresql.org/docs/current/sql-createpolicy.html)
- [Row Level Securityの変更](https://qiita.com/nuko_yokohama/items/c8d6d3f3f8f82e1051ec)
- [PostgreSQLのRow Level Securityを使ってマルチテナントデータを安全に扱う](https://times.hrbrain.co.jp/entry/postgresql-row-level-security)

### Others

- [Toggle between multiple .env files like .env.development with node.js](https://stackoverflow.com/questions/55406055/toggle-between-multiple-env-files-like-env-development-with-node-js)
- [Comparison of PostgreSQL and MongoDB](https://youtu.be/eM7hzKwvTq8)
- [Build your first Slack Bolt App Pt 3: OAuth & Storing User Credentials (w/ MongoDB)](https://youtu.be/sSfHz7j89Xk)
- [Bolt and MongoDB Sample](https://github.com/horeaporutiu/oauth-sample)

### CODE REVIEW

File: src/db/getInstall.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { supabase } from "../config/supabaseClient";
+
+// Get installation data
+// See https://supabase.com/docs/reference/javascript/select
+const getInstall = async (id: string) => {
+  const { data, error } = await supabase
```
Comment (huitzlopochtli) [inline src/db/getInstall.ts:6]: isn't there any way to execute raw SQL statements?

**Review by huitzlopochtli — APPROVED** [review]

File: src/db/getInstall.ts (hunk 1 above)
Comment (hiroshinishio) [reply src/db/getInstall.ts:7]: @huitzlopochtli told me 

- About the PR, implementation should not be bound to superbase library
- Suppose in future you want to change to MySQL or Postgres then the whole code will fail 
- In simple terms removing dependencies

Comment (hiroshinishio) [reply src/db/getInstall.ts:7]: @huitzlopochtli Is it ok if I use a pull request for another refactoring? I would like to ask you if possible.

Comment (hiroshinishio) [reply src/db/getInstall.ts:7]: @huitzlopochtli If OK, I would like this comment to be "Resolve".

**Review by huitzlopochtli — APPROVED** [review]

**Pull request merged**